### PR TITLE
changed yaml structure and python generator update

### PR DESCRIPTION
**Pull request by shubham-shay-te**

### CODE REVIEW

File: core_lib_generator/config_collectors/cache.py [hunk 1]
```diff
@@ -1,32 +1,37 @@
...
+def generate_cache_template() -> list:
+    cache_list = []
 
-    if cache_type == CacheTypes.Memory.value:
-        print(f'Cache type {CacheTypes(cache_type).name}')
-        return {
-            'config': {
-                cache_name: {
+    add_cache = True
+    while add_cache:
```
Comment (shay-te) [inline core_lib_generator/config_collectors/cache.py:10]: are we also generating the cache handlers in the CoreLib class?

File: core_lib_generator/config_collectors/data_access.py [hunk 2]
```diff
@@ -4,52 +4,53 @@
...
+                    f'What is the name of the data access? (Database connection: {db_conn}, Entity: {entity_name})',
                     default,
                     False,
                     is_exists,
                 )
             )
             is_crud_soft_delete_token = False
             is_crud_soft_delete = False
-            if db_entities[db_conn][entity]['is_soft_delete'] and db_entities[db_conn][entity]['is_soft_delete_token']:
+            if entity['is_soft_delete'] and entity['is_soft_delete_token']:
```
Comment (shay-te) [inline core_lib_generator/config_collectors/data_access.py:35]: entity.get('is_soft_delete')./ .get('is_soft_delete_token') is much more safe and will not raise any exceptions

File: core_lib_generator/config_collectors/data_access.py [hunk 3]
```diff
@@ -4,52 +4,53 @@
...
                 is_crud_soft_delete_token = input_yes_no(
                     'Do you want to implement CRUD Soft Delete Token on your data access?', False
                 )
                 is_crud_soft_delete = is_crud_soft_delete_token
                 is_crud = is_crud_soft_delete_token
-            elif (
-                db_entities[db_conn][entity]['is_soft_delete']
-                and not db_entities[db_conn][entity]['is_soft_delete_token']
-            ):
+            elif entity['is_soft_delete'] and not entity['is_soft_delete_token']:
```
Comment (shay-te) [inline core_lib_generator/config_collectors/data_access.py:41]: same. use get

File: core_lib_generator/config_collectors/db_entity.py [hunk 4]
```diff
@@ -3,15 +3,16 @@
...
     for db_conn in db:
-        if db[db_conn]['config']['url']['protocol'] == 'mongodb':
-            continue
-        entities.setdefault(db_conn, {})
+        # entities.setdefault(db_conn, {})
 
         def is_exists(user_input: str):
-            return False if user_input in entities[db_conn] else True
+            for entity in entities:
+                if entity['key'] == user_input and entity['db_connection'] == db_conn:
```
Comment (shay-te) [inline core_lib_generator/config_collectors/db_entity.py:13]: user `dict.get`

File: core_lib_generator/config_collectors/db_entity.py [hunk 5]
```diff
@@ -20,41 +21,52 @@ def is_exists(user_input: str):
...
-                elif column_type == DBDatatypes.VARCHAR.value:
-                    column_default = input_str(f'Enter the default value of column', '', True)
-                else:
-                    column_default = input_bool(
-                        f'Enter the default value of column (true, false, 0(false), 1(true))', 'true'
+            columns = []
+            if db[db_conn]['config']['url']['protocol'] != 'mongodb':
+                add_columns = input_yes_no(f'Do you want to add columns to `{entity_name}` entity?', True)
+
+                def is_exists_column(user_input: str):
```
Comment (shay-te) [inline core_lib_generator/config_collectors/db_entity.py:28]: can we use the generic function mention before?

File: core_lib_generator/core_lib_config_generate_yaml.py [hunk 6]
```diff
@@ -60,10 +64,11 @@ def _get_solr_config():
...
     print('Please fill out the requested Cache information.')
-    cache = generate_cache_template()
-    if 'env' in cache:
-        env.update(_get_env_variables(cache))
-    config['cache'] = cache['config']
+    cache_list = generate_cache_template()
+    for cache in cache_list:
+        if 'env' in cache:
+            env.update(_get_env_variables(cache))
+        config['cache'].append(cache['config'])
```
Comment (shay-te) [inline core_lib_generator/core_lib_config_generate_yaml.py:71]: config?

**Comment by shay-te** [comment]
Tomorrow we will make a demo for this

File: core_lib_generator/config_collectors/cache.py [hunk 7]
```diff
@@ -36,34 +41,32 @@ def _generate_cache_config(
     if cache_type == CacheTypes.Redis.value:
         return {
             'env': {
-                f'{cache_type_name.upper()}_PORT': cache_port,
-                f'{cache_type_name.upper()}_HOST': cache_host,
+                f'{cache_name.upper()}_PORT': cache_port,
+                f'{cache_name.upper()}_HOST': cache_host,
             },
             'config': {
```
[the remark this answers is not in the document]
Comment (shubham-shay-te) [reply core_lib_generator/config_collectors/cache.py:47]: for the separation between `env` and the main config `section`

File: core_lib_generator/config_collectors/cache.py [hunk 8]
```diff
@@ -36,34 +41,32 @@ def _generate_cache_config(
...
         }
     else:
         return {
             'env': {
-                f'{cache_type_name.upper()}_PORT': cache_port,
-                f'{cache_type_name.upper()}_HOST': cache_host,
+                f'{cache_name.upper()}_PORT': cache_port,
+                f'{cache_name.upper()}_HOST': cache_host,
             },
             'config': {
```
[the remark this answers is not in the document]
Comment (shubham-shay-te) [reply core_lib_generator/config_collectors/cache.py:63]: for the separation between `env` and the main config `section`

File: core_lib_generator/config_collectors/cache.py (hunk 1 above)
Comment (shubham-shay-te) [reply core_lib_generator/config_collectors/cache.py:10]:
```python
CoreLib.cache_registry.register('memorycache', CacheHandlerRam())
        CoreLib.cache_registry.register('rediscache', CacheHandlerRedis(build_url(**self.config.core_lib.example_core_lib.rediscache.cache.url)))
```
we are generating this

Comment (shay-te) [reply core_lib_generator/config_collectors/cache.py:10]: Good work!

File: core_lib_generator/core_lib_config_generate_yaml.py (hunk 6 above)
Comment (shubham-shay-te) [reply core_lib_generator/core_lib_config_generate_yaml.py:71]: we will have 2 sections in the returned dict `env` and `config` we will have to append the config section into the final list so for that reason

File: core_lib_generator/config_collectors/cache.py (hunk 1 above)
Comment (shubham-shay-te) [reply core_lib_generator/config_collectors/cache.py:17]: thankyou

File: core_lib_generator/config_collectors/cache.py [hunk 9]
```diff
@@ -1,32 +1,44 @@
...
-        cache_type_name = CacheTypes(cache_type).name
-        cache_port = input_int(f'Enter your {cache_type_name} server port no.', default_cache_ports[cache_type_name])
-        cache_host = input_str(f'Enter your {cache_type_name} server host', 'localhost')
-        cache_protocol = None
-        if cache_type == CacheTypes.Redis.value:
-            cache_protocol = input_str(f'Enter your {cache_type_name} protocol', f'{cache_type_name.lower()}')
-        print(f'Cache type {cache_type_name} on {cache_host}:{cache_port}')
-        return _generate_cache_config(cache_name, cache_type, cache_port, cache_host, cache_protocol)
+            })
+        elif cache_type == CacheTypes.Empty.value:
```
Comment (shay-te) [inline core_lib_generator/config_collectors/cache.py:29]: remove empty part

File: core_lib_generator/config_collectors/cache.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (shay-te) [reply core_lib_generator/config_collectors/cache.py:70]: config -> cache

File: core_lib_generator/config_collectors/db_entity.py [hunk 10]
```diff
@@ -1,60 +1,69 @@
...
-                if column_type == DBDatatypes.INTEGER.value:
-                    column_default = input_int(f'Enter the default value of column', 0)
-                elif column_type == DBDatatypes.VARCHAR.value:
-                    column_default = input_str(f'Enter the default value of column', '', True)
-                else:
-                    column_default = input_bool(
-                        f'Enter the default value of column (true, false, 0(false), 1(true))', 'true'
+            columns = []
+            if get_dict_attr(db, f'{db_conn}.config.url.protocol') != 'mongodb':
+                add_columns = input_yes_no(f'Do you want to add columns to `{entity_name}` entity?', True)
```
Comment (shay-te) [inline core_lib_generator/config_collectors/db_entity.py:26]: remove add for columns in the mongo part

**Review by shay-te — APPROVED** [review]

**Pull request merged**